odin/src/main/kotlin/me/odin/features/impl/floor7/p3/ArrowsDevice.kt

```diff
@@ -0,0 +1,193 @@
+import me.odinmain.utils.render.Renderer
+import me.odinmain.utils.skyblock.*
+import me.odinmain.utils.skyblock.dungeon.DungeonUtils
+import me.odinmain.utils.skyblock.dungeon.M7Phases
+import net.minecraft.entity.item.EntityArmorStand
+import net.minecraft.init.Blocks
+import net.minecraft.util.AxisAlignedBB
+import net.minecraft.util.BlockPos
+import net.minecraftforge.client.event.RenderWorldLastEvent
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+import net.minecraftforge.fml.common.gameevent.TickEvent.ClientTickEvent
+import org.lwjgl.input.Keyboard
+
+object ArrowsDevice : Module(
+    name = "Arrows Device",
+    description = "Solver for the Sharp Shooter puzzle in floor 7.",
+    category = Category.FLOOR7,
+) {
+    private val solver: Boolean by BooleanSetting("Solver")
+    private val markedPositionColor: Color by ColorSetting("Marked Position", Color.RED, description = "Color of the marked position.").withDependency { solver }
+    private val targetPositionColor: Color by ColorSetting("Target Position", Color.GREEN, description = "Color of the target position.").withDependency { solver }
+    private val alertOnDeviceComplete: Boolean by BooleanSetting("Device complete alert", default = true, description = "Send an alert when device is complete")
+    private val resetKey: Keybinding by KeybindSetting("Reset", Keyboard.KEY_NONE, description = "Resets the solver.").onPress {
+        reset()
+    }.withDependency { solver }
+    private val depthCheck: Boolean by BooleanSetting("Depth check", true, description = "Marked positions show through walls.").withDependency { solver }
+    private val reset: () -> Unit by ActionSetting("Reset") {
+        markedPositions.clear()
+    }.withDependency { solver }
+
+    private val markedPositions = mutableSetOf<BlockPos>()
+    private var targetPosition: BlockPos? = null
+
+    private var isDeviceComplete = false
+
+    // ArmorStand showing the status of the device
+    private var activeArmorStand: EntityArmorStand? = null
+
+    // Number of server ticks since the last target disappeared, or null if there is a target
+    private var serverTicksSinceLastTargetDisappeared: Int? = null
+
+    init {
+        onMessage(Regex("^[a-zA-Z0-9_]{3,} completed a device! \\([1-7]/7\\)"), { enabled && isPlayerInRoom }) {
+            onComplete()
+        }
+
+        onMessage(Regex("^ ☠ You died and became a ghost\\.$"), { enabled && isPlayerOnStand }) {
+            // Prevent the tick count from continuing and registering a false positive
+            // This does mean we could get a false negative but since the player is already
+            // dead knowing if the device is complete has less importance (no leaping)
+            serverTicksSinceLastTargetDisappeared = 11
+        }
+
+        execute(1000) {
+            if (DungeonUtils.getPhase() != M7Phases.P3) return@execute
+
+            // Cast is safe since we won't return an entity that isn't an armor stand
```

Comment:
you can just filterIsInstance<EntityArmorStand>() before your find and you will not need to cast